aep/general/0155/aep.md

```diff
@@ -0,0 +1,106 @@
+# Idempotency
+
+It is sometimes useful for an API to have a unique, customer-provided
+identifier for particular requests. This can be useful for several purposes,
+such as de-duplicating requests from parallel processes, ensuring the safety of
+retries, or auditing.
+
+The purpose of idempotency keys is to provide idempotency guarantees: allowing
+the same request to be issued more than once without subsequent calls having
+any effect. In the event of a network failure, the client can retry the
+request, and the server can detect duplication and ensure that the request is
+only processed once.
+
+## Guidance
+
+APIs **may** add a `string idempotency_key` parameter to request messages
+(including those of standard methods) in order to uniquely identify particular
+requests.
+
+```proto
+message CreateBookRequest {
+  // The parent resource where this book will be created.
+  // Format: publishers/{publisher}
+  string parent = 1 [
+    (google.api.field_behavior) = REQUIRED,
+    (google.api.resource_reference) = {
+      child_type: "library.example.com/Book"
+    }];
+
+  // The book to create.
+  Book book = 2 [(google.api.field_behavior) = REQUIRED];
+
+  // A unique identifier for this request. Restricted to 36 ASCII characters.
+  // A random UUID is recommended.
+  // This request is only idempotent if a `idempotency_key` is provided.
+  string idempotency_key = 3 [(google.api.field_info).format = UUID4];
+}
+```
+
+- Providing an idempotency key **must** guarantee idempotency.
+  - If a duplicate request is detected, the server **should** return the
+    response for the previously successful request, because the client most
+    likely did not receive the previous response.
+  - APIs **may** choose any reasonable timeframe for honoring idempotency keys.
```

Comment:
is there value in defining how one can be aware of what the timeframe is? As a laymans user who didn't read this AIP, I'd imagine idempotency_keys are eternal, and I should never be able to do an action twice.

e.g. look at Amazon's design pattern on this: https://aws.amazon.com/builders-library/making-retries-safe-with-idempotent-APIs/. They also talk about labeling the resource so a user can find the created resource easily.

Reply:
From meeting (myself, @toumorokoshi, Andrei):

* Putting the key on the resource isn't a comprehensive solution (e.g. stateless methods have no resource) and so maybe isn't the right path.
* This guidance allows e.g. a 100ms lifetime on an idempotency key, which would make it useless.  How are users suppose to interpret and trust this pattern?  What is the purpose of this field?
* At a minimum, we should guide that APIs **must** document the minimum lifetime of an idempotency key.
  * Maybe we introduce a field annotation for this.  Under the `aep.api` package, since we can't block on Google letting us add stuff to `google.api`?
    * We aren't going to fork existing `google.api` annotations, though, because e.g. gRPC/JSON transcoding implementations (like .NET 7's) depend on them.
    * What does the long-term AEP alignment plan look like for Google?  Anything stopping them using `aep.*` protos?  (Or they could duplicate the types to `google.api`, or they could have a gateway that translates, ...)

Tentative conclusions:

* Create an annotation in the `aep` repo, then link to it here and mandate its usage.
* Provide **should**-level guidance about minimum `idempotency_key` lifetime.

Reply:
Re the second bullet, I've updated the wording of the guidance. 

Re the first, I've created https://github.com/aep-dev/aep/pull/6.  I'd like to get that reviewed and merged before updating this.

Reply:
I don't see the update on a recommended minimum lifetime, can you point me to it?

Reply:
Oops, I forgot to push my changes.  Now pushed.